File: js/web/lib/wasm/jsep/webgpu/ops/3rd-party/conv_backprop_webgpu.ts
/**
 * @license
 * Copyright 2021 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

// sampled from [@tensorflow/tfjs] tfjs-backend-webgpu/src/conv_backprop_webgpu.ts

import { DataType } from '../../../../wasm-common';
import { LOG_DEBUG } from '../../../log';
import { TensorView } from '../../../tensor-view';
import { ShapeUtil } from '../../../util';
import { ProgramInfo, ProgramInputTensorInfoDependency, ProgramUniform } from '../../types';
import {
  createTensorShapeVariables,
  inputVariable,
  outputVariable,
  ShaderHelper,
  tensorTypeToWsglStorageType,
  UniformsArrayType,
  getMaxComponents,
} from '../common';
import { ConvTransposeAttributes } from '../conv-transpose';

export const createConvTranspose2DProgramInfo = (
  inputs: readonly TensorView[],
  attributes: ConvTransposeAttributes,
  squeezeOutputShapeFunction?: (shape: readonly number[]) => number[],
): ProgramInfo => {
  const hasBias = inputs.length > 2;
  const outputShape = attributes.outputShape;
  const isChannelsLast = attributes.format === 'NHWC';
  const group = attributes.group;
  const wShape = inputs[1].dims;
  const inputChannelsPerGroup = wShape[2] / group;
  const outputChannelsPerGroup = wShape[3];
  const aComponents = isChannelsLast ? getMaxComponents(inputChannelsPerGroup) : 1;
  const components = isChannelsLast ? getMaxComponents(outputChannelsPerGroup) : 1;
  const bComponents = isChannelsLast ? (outputChannelsPerGroup === 1 ? aComponents : components) : 1;
  const outputSize = ShapeUtil.size(outputShape) / components;
  const dispatch = [Math.ceil(outputSize / 64), 1, 1];
  LOG_DEBUG('verbose', () => `[conv2d_backprop_webgpu] dispatch = ${dispatch}`);

  const inputDependencies: ProgramInputTensorInfoDependency[] = ['rank', 'rank'];
  const strides = [attributes.strides[0], attributes.strides[1]];
  const filterDims = [attributes.kernelShape[isChannelsLast ? 1 : 2], attributes.kernelShape[isChannelsLast ? 2 : 3]];
  const dilations = [attributes.dilations[0], attributes.dilations[1]];
  const effectiveFilterDims = [
    filterDims[0] +
      (attributes.dilations[0] <= 1
        ? 0
        : (attributes.kernelShape[isChannelsLast ? 1 : 2] - 1) * (attributes.dilations[0] - 1)),
    filterDims[1] +
      (attributes.dilations[1] <= 1
        ? 0
        : (attributes.kernelShape[isChannelsLast ? 2 : 3] - 1) * (attributes.dilations[1] - 1)),
  ];
  const pads = [
    effectiveFilterDims[0] - 1 - Math.floor((attributes.pads[0] + attributes.pads[2]) / 2),
    effectiveFilterDims[1] - 1 - Math.floor((attributes.pads[1] + attributes.pads[3]) / 2),
  ];

  const programUniforms: ProgramUniform[] = [
    { type: DataType.uint32, data: outputSize },
    { type: DataType.uint32, data: strides },
    { type: DataType.uint32, data: filterDims },
    { type: DataType.uint32, data: dilations },
    { type: DataType.uint32, data: effectiveFilterDims },
    { type: DataType.int32, data: pads },
    { type: DataType.uint32, data: inputChannelsPerGroup },
    { type: DataType.uint32, data: outputChannelsPerGroup },
    ...createTensorShapeVariables(inputs[0].dims, inputs[1].dims),
  ];
  if (hasBias) {
    programUniforms.push(...createTensorShapeVariables(inputs[2].dims));
    inputDependencies.push('rank');
  }
  programUniforms.push(...createTensorShapeVariables(outputShape));

  const getShaderSource = (shaderHelper: ShaderHelper) => {
    const uniforms: UniformsArrayType = [
      { name: 'output_size', type: 'u32' },
      { name: 'strides', type: 'u32', length: strides.length },
      { name: 'filter_dims', type: 'u32', length: filterDims.length },
      { name: 'dilations', type: 'u32', length: filterDims.length },
      { name: 'effective_filter_dims', type: 'u32', length: effectiveFilterDims.length },
      { name: 'pads', type: 'i32', length: pads.length },
      { name: 'input_channels_per_group', type: 'u32' },
      { name: 'output_channels_per_group', type: 'u32' },
    ];
    const dataType = tensorTypeToWsglStorageType(inputs[0].dataType);
    const rowDim = isChannelsLast ? 1 : 2;
    const colDim = isChannelsLast ? 2 : 3;
    const channelDim = isChannelsLast ? 3 : 1;

    const w = inputVariable('W', inputs[1].dataType, inputs[1].dims.length, bComponents);
    const dy = inputVariable('Dy', inputs[0].dataType, inputs[0].dims.length, aComponents);
    const inputVariables = [dy, w];
    if (hasBias) {
      inputVariables.push(inputVariable('bias', inputs[2].dataType, [outputShape[channelDim]].length, components));
    }
    const output = outputVariable('result', inputs[0].dataType, outputShape.length, components);

    const calculateResult = (): string => {
      let calcStr = '';
      if (aComponents === 1) {
        calcStr += `
        let w_offset = ${w.indicesToOffset(`${w.type.indices}(u32(wRPerm), u32(wCPerm), inputChannel, wOutChannel)`)};
        let wValue = ${w.getByOffset(`w_offset / ${bComponents}`)};
        dotProd = dotProd + xValue * wValue;`;
      } else {
        if (outputChannelsPerGroup === 1) {
          calcStr += `
          let wValue = ${w.getByOffset(`${w.indicesToOffset(`${w.type.indices}(u32(wRPerm), u32(wCPerm), inputChannel, wOutChannel)`)} / ${bComponents}`)};
          dotProd = dotProd + dot(xValue, wValue);`;
        } else {
          for (let c = 0; c < aComponents; c++) {
            calcStr += `
            let wValue${c} = ${w.getByOffset(`${w.indicesToOffset(`${w.type.indices}(u32(wRPerm), u32(wCPerm), inputChannel + ${c}, wOutChannel)`)} / ${bComponents}`)};
            dotProd = dotProd + xValue[${c}] * wValue${c};`;
          }
        }
      }
      return calcStr;
    };
    const codeSnippet = `
            let outputIndices = ${output.offsetToIndices(`global_idx * ${components}`)};
            let batch = ${output.indicesGet('outputIndices', 0)};
            let d1 = ${output.indicesGet('outputIndices', channelDim)};
            let r = ${output.indicesGet('outputIndices', rowDim)};
            let c = ${output.indicesGet('outputIndices', colDim)};
            let dyCorner = vec2<i32>(i32(r), i32(c)) - uniforms.pads;
            let dyRCorner = dyCorner.x;
            let dyCCorner = dyCorner.y;
            let groupId = d1 / uniforms.output_channels_per_group;
            let wOutChannel = d1 - groupId * uniforms.output_channels_per_group;
            // Convolve dy(?, ?, d2) with w(:, :, d1, d2) to compute dx(xR, xC, d1).
            // ? = to be determined. : = across all values in that axis.
            var dotProd = ${output.type.value}(0.0);
            var wR: u32 = 0;
            if (uniforms.dilations.x == 1) {
              // Minimum wR >= 0 that satisfies (dyRCorner + wR) % (uniforms.strides.x) == 0
              wR = u32(((dyRCorner + i32(uniforms.strides.x) - 1) / i32(uniforms.strides.x)) * i32(uniforms.strides.x) - dyRCorner);
            }
            for (; wR < uniforms.effective_filter_dims.x; wR = wR + 1) {
              if (wR % uniforms.dilations.x != 0) {
                continue;
              }
              let dyR = (${dataType}(dyRCorner) + ${dataType}(wR)) / ${dataType}(uniforms.strides[0]);
              let wRPerm = uniforms.filter_dims.x - 1 - wR / uniforms.dilations.x;
              if (dyR < 0.0 || dyR >= ${dataType}(uniforms.Dy_shape[${rowDim}]) || fract(dyR) > 0.0 ||
                  wRPerm < 0) {
                continue;
              }
              let idyR: u32 = u32(dyR);
              var wC: u32 = 0;
              if (uniforms.dilations.y == 1) {
                // Minimum wC >= 0 that satisfies (dyCCorner + wC) % (uniforms.strides.y) == 0
                wC = u32(((dyCCorner + i32(uniforms.strides.y) - 1) / i32(uniforms.strides.y)) * i32(uniforms.strides.y) - dyCCorner);
              }

              for (; wC < uniforms.effective_filter_dims.y; wC = wC + 1) {
                if (wC % uniforms.dilations.y != 0) {
                  continue;
                }
                let dyC = (${dataType}(dyCCorner) + ${dataType}(wC)) / ${dataType}(uniforms.strides.y);
                let wCPerm = uniforms.filter_dims.y - 1 - wC / uniforms.dilations.y;
                if (dyC < 0.0 || dyC >= ${dataType}(uniforms.Dy_shape[${colDim}]) ||
                    fract(dyC) > 0.0 || wCPerm < 0) {
                  continue;
                }
                let idyC: u32 = u32(dyC);
                var inputChannel = groupId * uniforms.input_channels_per_group;
                for (var d2: u32 = 0; d2 < uniforms.input_channels_per_group; d2 = d2 + ${aComponents}) {
                  let xValue = ${
                    isChannelsLast
                      ? dy.getByOffset(
                          `${dy.indicesToOffset(`${dy.type.indices}(batch, idyR, idyC, inputChannel)`)} / ${aComponents}`,
                        )
                      : dy.get('batch', 'inputChannel', 'idyR', 'idyC')
                  };
                  ${calculateResult()}
                  inputChannel = inputChannel + ${aComponents};
                }
                wC = wC + uniforms.strides.y - 1;
              }
              wR = wR + uniforms.strides[0] - 1;
            }
            let value = dotProd${hasBias ? ` + bias[d1 / ${components}]` : ''};
            ${output.setByOffset('global_idx', 'value')};
          `;

    return `
    ${shaderHelper.registerUniforms(uniforms).declareVariables(...inputVariables, output)}
      ${shaderHelper.mainStart()}
      ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes('uniforms.output_size')};
    ${codeSnippet}}`;
  };

  return {
    name: 'ConvTranspose2D',
    shaderCache: {
      hint: `${attributes.cacheKey};${aComponents}${bComponents}${components}${outputChannelsPerGroup === 1}`,
      inputDependencies,
    },
    getRunData: () => ({
      dispatchGroup: { x: dispatch[0], y: dispatch[1], z: dispatch[2] },
      outputs: [
        {
          dims: squeezeOutputShapeFunction ? squeezeOutputShapeFunction(outputShape) : outputShape,
          dataType: inputs[0].dataType,
        },
      ],
      programUniforms,
    }),
    getShaderSource,
  };
};

File: onnxruntime/core/providers/webgpu/nn/conv_backprop_webgpu.cc
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
#include <vector>
#include <string>
#include <sstream>
#include "core/common/inlined_containers.h"
#include "core/providers/webgpu/nn/conv_backprop_webgpu.h"
#include "core/providers/webgpu/webgpu_utils.h"
namespace onnxruntime {
namespace webgpu {

Status ConvTranspose2DProgram::GenerateShaderCode(ShaderHelper& shader) const {
  const auto& dy = shader.AddInput("dy", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseValueTypeAlias | ShaderUsage::UseElementTypeAlias);
  const auto& w = shader.AddInput("w", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseValueTypeAlias);
  const auto& output = shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseValueTypeAlias);
  if (has_bias_) {
    shader.AddInput("bias");
  }
  auto row_dim = is_channels_last_ ? 1 : 2;
  auto col_dim = is_channels_last_ ? 2 : 3;
  auto channel_dim = is_channels_last_ ? 3 : 1;
  auto calculate_result = [&]() -> std::string {
    std::stringstream ss;
    if (pack_input_as4_) {
      if (a_components_ == 4) {
        ss << "let xValue = " << dy.GetByOffset("x_offset") << ";\n"
           << "let wValue = " << w.GetByOffset("w_offset") << ";\n"
           << "dotProd = dotProd + dot(xValue, wValue);\n"
           << "x_offset += 1;\n"
           << "w_offset += 1;\n";
      } else if (a_components_ == 2) {
        ss << "let xValue = vec4<dy_element_t>(" << dy.GetByOffset("x_offset") << ", " << dy.GetByOffset("x_offset + 1") << ");\n"
           << "let wValue = vec4<dy_element_t>(" << w.GetByOffset("w_offset") << ", " << w.GetByOffset("w_offset + 1u") << ");\n"
           << "dotProd = dotProd + dot(xValue, wValue);\n"
           << "x_offset += 2;\n"
           << "w_offset += 2;\n";
      } else if (a_components_ == 1) {
        ss << "let xValue = vec4<dy_element_t>(" << dy.GetByOffset("x_offset") << ", " << dy.GetByOffset("x_offset + 1u") << ", " << dy.GetByOffset("x_offset + 2u") << ", " << dy.GetByOffset("x_offset + 3u") << ");\n"
           << "let wValue = vec4<dy_element_t>(" << w.GetByOffset("x_offset") << ", " << w.GetByOffset("x_offset + 1u") << ", " << w.GetByOffset("x_offset + 2u") << ", " << w.GetByOffset("x_offset + 3u") << ");\n"
           << "dotProd = dotProd + dot(xValue, wValue);\n"
           << "x_offset += 4;\n"
           << "w_offset += 4;\n";
      }
    } else {
      if (is_channels_last_) {
        ss << "let xValue = " << dy.GetByIndices("dy_indices_t(batch, idyR, idyC, inputChannel / " + std::to_string(a_components_)) << ");\n";
      } else {
        ss << "let xValue = " << dy.GetByIndices("dy_indices_t(batch, inputChannel, idyR, idyC)") << ";\n";
      }
      if (a_components_ == 1) {
        ss << "let wValue = " << w.GetByIndices("w_indices_t(u32(wRPerm), u32(wCPerm), inputChannel, wOutChannel / " + std::to_string(b_components_) + ")") << ";\n"
           << "dotProd = dotProd + xValue * wValue;\n";
      } else if (a_components_ == b_components_ && components_ == 1) {
        ss << "let wValue = " << w.GetByIndices("w_indices_t(u32(wRPerm), u32(wCPerm), inputChannel, wOutChannel)") << ";\n"
           << "dotProd = dotProd + dot(xValue, wValue);\n";
      } else {
        for (uint32_t i = 0; i < a_components_; ++i) {
          ss << "let w_indices" << i << " = w_indices_t(u32(wRPerm), u32(wCPerm), inputChannel + " << i << ", wOutChannel / " << b_components_ << ");\n "
             << "let wValue" << i << " = " << w.GetByIndices("w_indices" + std::to_string(i)) << ";\n"
             << "dotProd = dotProd + xValue[" << i << "] * wValue" << i << ";\n";
        }
      }
    }
    return ss.str();
  };
  auto calculate_remainder = [&]() -> std::string {
    std::stringstream ss;
    if (input_channels_remainder_ > 0) {
      ORT_ENFORCE(pack_input_as4_, "Invalid input_channels_remainder: ", input_channels_remainder_);
      if (a_components_ == 1) {
        for (uint32_t i = 0; i < input_channels_remainder_; ++i) {
          ss << "dotProd = dotProd + " << dy.GetByOffset("x_offset + " + std::to_string(i)) << ";\n";
        }
      } else if (a_components_ == 2) {
        if (input_channels_remainder_ != 2) {
          ORT_THROW("Invalid input_channels_remainder: ", input_channels_remainder_);
        }
        ss << "let xValue = " << dy.GetByOffset("x_offset") << ";\n"
           << "let wValue = " << w.GetByOffset("w_offset") << ";\n"
           << "dotProd = dotProd + dot(xValue, wValue);\n";
      }
    }
    return ss.str();
  };
  shader.MainFunctionBody() << shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size")
                            << "let outputIndices = " << output.OffsetToIndices("global_idx") << ";\n"
                            << "let batch = " << output.IndicesGet("outputIndices", 0) << ";\n"
                            << "let d1 = " << output.IndicesGet("outputIndices", channel_dim) << " * " << components_ << ";\n"
                            << "let r = " << output.IndicesGet("outputIndices", row_dim) << ";\n"
                            << "let c = " << output.IndicesGet("outputIndices", col_dim) << ";\n"
                            << "let dyCorner = vec2<i32>(i32(r), i32(c)) - vec2<i32>(uniforms.pads);\n"
                            << "let dyRCorner = dyCorner.x;\n"
                            << "let dyCCorner = dyCorner.y;\n"
                            << "let groupId = d1 / uniforms.output_channels_per_group;\n"
                            << "let wOutChannel = d1 - groupId * uniforms.output_channels_per_group;\n"
                            << "// Convolve dy(?, ?, d2) with w(:, :, d1, d2) to compute dx(xR, xC, d1).\n"
                            << "// ? = to be determined. : = across all values in that axis.\n"
                            << "var dotProd = output_value_t(0.0);\n"
                            << "var wR: u32 = 0;\n"
                            << "if (uniforms.dilations.x == 1) {\n"
                            << "  // Minimum wR >= 0 that satisfies (dyRCorner + wR) % (uniforms.strides.x) == 0\n"
                            << "  wR = u32(((dyRCorner + i32(uniforms.strides.x) - 1) / i32(uniforms.strides.x)) * i32(uniforms.strides.x) - dyRCorner);\n"
                            << "}\n"
                            << "for (; wR < uniforms.effective_filter_dims.x; wR = wR + 1) {\n"
                            << "  if (wR % uniforms.dilations.x != 0) {\n"
                            << "    continue;\n"
                            << "  }\n"
                            << "  let dyR = (dy_element_t(dyRCorner) + dy_element_t(wR)) / dy_element_t(uniforms.strides[0]);\n"
                            << "  let wRPerm = uniforms.filter_dims.x - 1 - wR / uniforms.dilations.x;\n"
                            << "  if (dyR < 0.0 || dyR >= dy_element_t(uniforms.dy_shape[" << row_dim << "]) || fract(dyR) > 0.0 || wRPerm < 0) {\n"
                            << "    continue;\n"
                            << "  }\n"
                            << "  let idyR: u32 = u32(dyR);\n"
                            << "  var wC: u32 = 0;\n"
                            << "  if (uniforms.dilations.y == 1) {\n"
                            << "    // Minimum wC >= 0 that satisfies (dyCCorner + wC) % (uniforms.strides.y) == 0\n"
                            << "    wC = u32(((dyCCorner + i32(uniforms.strides.y) - 1) / i32(uniforms.strides.y)) * i32(uniforms.strides.y) - dyCCorner);\n"
                            << "  }\n"
                            << "  for (; wC < uniforms.effective_filter_dims.y; wC = wC + 1) {\n"
                            << "    if (wC % uniforms.dilations.y != 0) {"
                            << "      continue;\n"
                            << "    }\n"
                            << "    let dyC = (dy_element_t(dyCCorner) + dy_element_t(wC)) / dy_element_t(uniforms.strides.y);\n"
                            << "    let wCPerm = uniforms.filter_dims.y - 1 - wC / uniforms.dilations.y;\n"
                            << "    if (dyC < 0.0 || dyC >= dy_element_t(uniforms.dy_shape[" << col_dim << "]) ||\n"
                            << "        fract(dyC) > 0.0 || wCPerm < 0) {\n"
                            << "      continue;\n"
                            << "    }\n"
                            << "    let idyC: u32 = u32(dyC);\n"
                            << "    var inputChannel = groupId * uniforms.input_channels_per_group;\n";
  if (pack_input_as4_) {
    shader.MainFunctionBody() << "    let dy_indices = dy_indices_t(batch, idyR, idyC, inputChannel / " << a_components_ << ");\n"
                              << "    let w_indices = w_indices_t(u32(wRPerm), u32(wCPerm), inputChannel, wOutChannel / " << b_components_ << ");\n"
                              << "    var x_offset = " << dy.IndicesToOffset("dy_indices") << ";\n"
                              << "    var w_offset = " << w.IndicesToOffset("w_indices") << ";\n";
  }

  shader.MainFunctionBody() << "    for (var d2: u32 = 0; d2 < uniforms.input_channels_per_group_int; d2 = d2 + " << (pack_input_as4_ ? 4 : a_components_) << ") {\n"
                            << "      " << calculate_result() << "\n"
                            << "      inputChannel = inputChannel + " << (pack_input_as4_ ? 4 : a_components_) << ";\n"
                            << "    }\n"
                            << "    " << calculate_remainder() << "\n"
                            << "    wC = wC + uniforms.strides.y - 1;\n"
                            << "  }\n"
                            << "  wR = wR + uniforms.strides.x - 1;\n"
                            << "}\n"
                            << "let value = dotProd" << (has_bias_ ? " + bias[d1 / " + std::to_string(components_) + "]" : "") << ";\n"
                            << output.SetByOffset("global_idx", "value") << "\n";
  return Status::OK();
}

ConvTranspose2DProgram CreateConvTranspose2DProgram(const std::vector<const Tensor*>& inputs, const std::vector<uint32_t>& pads, const std::vector<uint32_t>& strides, const std::vector<uint32_t>& dilations, Tensor* output, bool is_channels_last, const std::vector<TensorShape>& modified_input_output_shapes, uint32_t groups) {
  bool has_bias = inputs.size() > 2;
  const auto* input = inputs[0];
  const auto* weight = inputs[1];
  const auto& input_shape = modified_input_output_shapes[0];
  const auto& weight_shape = modified_input_output_shapes[1];
  const auto& output_shape = modified_input_output_shapes[has_bias ? 3 : 2];
  auto input_channels_per_group = weight_shape[2] / groups;
  auto output_channels_per_group = weight_shape[3];
  auto a_components = is_channels_last ? GetMaxComponents(input_channels_per_group) : 1;
  bool pack_input_as4 = is_channels_last && output_channels_per_group == 1 && input_channels_per_group >= 4;
  auto input_channels_per_group_int = pack_input_as4 ? (input_channels_per_group / 4) * 4 : (input_channels_per_group / a_components) * a_components;
  auto input_channels_remainder = input_channels_per_group - input_channels_per_group_int;
  auto components = is_channels_last ? GetMaxComponents(output_channels_per_group) : 1;
  auto b_components = is_channels_last ? (output_channels_per_group == 1 ? a_components : components) : 1;
  TensorShape reduced_input_shape = ReduceShapeByComponents(input_shape, a_components);
  TensorShape reduced_weight_shape = ReduceShapeByComponents(weight_shape, b_components);
  TensorShape reduced_output_shape = ReduceShapeByComponents(output_shape, components);
  auto output_size = reduced_output_shape.Size();
  std::vector<uint32_t> kernel_dims = {static_cast<uint32_t>(weight_shape[0]), static_cast<uint32_t>(weight_shape[1])};
  std::vector<uint32_t> effective_kernel_dims = {kernel_dims[0] + ((dilations[0] <= 1) ? 0 : ((kernel_dims[0] - 1) * (dilations[0] - 1))), kernel_dims[1] + ((dilations[1] <= 1) ? 0 : ((kernel_dims[1] - 1) * (dilations[1] - 1)))};
  std::vector<uint32_t> local_pads = {effective_kernel_dims[0] - 1 - pads[0], effective_kernel_dims[1] - 1 - pads[1]};
  ConvTranspose2DProgram program(is_channels_last, has_bias, components, a_components, b_components, uint32_t(input_channels_remainder), pack_input_as4);
  program.AddInputs({{input, ProgramTensorMetadataDependency::TypeAndRank, reduced_input_shape, a_components}, {weight, ProgramTensorMetadataDependency::TypeAndRank, reduced_weight_shape, b_components}});
  if (has_bias) {
    const auto* bias = inputs[2];
    const auto& bias_shape = modified_input_output_shapes[2];
    TensorShape reduced_bias_shape = ReduceShapeByComponents(bias_shape, components);
    program.AddInput({bias, ProgramTensorMetadataDependency::TypeAndRank, reduced_bias_shape, components});
  }
  program.AddOutput({output, ProgramTensorMetadataDependency::Rank, reduced_output_shape, components})
      .CacheHint(input_channels_remainder, pack_input_as4, components, b_components, a_components, is_channels_last ? 1 : 0, has_bias ? 1 : 0)
      .AddUniformVariables({{static_cast<uint32_t>(output_size)}, {strides}, {kernel_dims}, {dilations}, {effective_kernel_dims}, {local_pads}, {static_cast<uint32_t>(input_channels_per_group_int)}, {static_cast<uint32_t>(input_channels_per_group)}, {static_cast<uint32_t>(output_channels_per_group)}})
      .SetDispatchGroupSize((output_size + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE);

  return program;
}

}  // namespace webgpu
}  // namespace onnxruntime
